javascript/ql/src/experimental/Security/CWE-444/InsecureHttpParser.ql

```diff
@@ -0,0 +1,42 @@
+/**
+ * @name Insecure http parser
+ * @description Using an insecure http parser can lead to http smuggling attacks.
+ * @kind problem
+ * @problem.severity warning
+ * @security-severity 9.0
+ * @precision high
+ * @id js/insecure-http-parser
+ * @tags security
+ *       external/cwe/cwe-444
+ */
+
+import javascript
+
+/** Gets options argument for a potential http or https connection */
+DataFlow::InvokeNode nodeInvocation() {
+  result instanceof ClientRequest
+  or
+  result instanceof Http::ServerDefinition
+}
+
+/** Gets an options object for an http or https connection. */
+DataFlow::ObjectLiteralNode nodeOptions() { result.flowsTo(nodeInvocation().getAnArgument()) }
+
+from DataFlow::PropWrite disable
+where
+  exists(DataFlow::SourceNode env |
+    env = NodeJSLib::process().getAPropertyRead("env") and
+    disable = env.getAPropertyWrite("NODE_OPTIONS") and
+    disable.getRhs().getStringValue().matches("%--insecure-http-parser%")
```

Comment:
Here you are requiring that `disable.getRhs()` should be a certain string value.  
But in the last line of the `where` you require it to be the boolean value `true`.  
Those two can't be true at the same time.

Reply:
My understanding was that the `and` took precedence and thus the flow was treated like this: `where a1 Or (a2 And b)
`. I originally had parenthesis to guarantee this was the case, but the formatter removed them. It proved to produce the desired result in my testing as well. Let me know if I am misunderstanding what is going on.

Reply:
The parameter can exist as a server option `insecureHTTPParser=true` or in the `NODE_OPTIONS` as a cli parameter `NODE_OPTIONS="--insecure-http-parser"`. So I have to do both boolean and string comparison.

Reply:
Ahhh, right. I didn't notice how the parentheses are grouped.  
It's good 👍